packages/core/src/config/utils.ts

```diff
@@ -0,0 +1,188 @@
+    Object.assign(result.oas3_0Decorators, rulesConf.oas3_0Decorators);
+    assignExisting(result.oas3_0Decorators, rulesConf.decorators || {});
+    Object.assign(result.oas3_1Decorators, rulesConf.oas3_1Decorators);
+    assignExisting(result.oas3_1Decorators, rulesConf.decorators || {});
+
+    result.plugins!.push(...(rulesConf.plugins || []));
+    result.pluginPaths!.push(...(rulesConf.pluginPaths || []));
+    result.extendPaths!.push(...new Set(rulesConf.extendPaths));
+  }
+
+  return result;
+}
+
+export function getMergedConfig(config: Config, entrypointAlias?: string): Config {
+  return entrypointAlias
+    ? new Config(
+        {
+          ...config.rawConfig,
+          lint: config.apis[entrypointAlias]
+            ? config.apis[entrypointAlias]?.lint
+            : config.rawConfig.lint,
+          'features.openapi': {
+            ...config['features.openapi'],
+            ...config.apis[entrypointAlias]?.['features.openapi'],
+          },
+          'features.mockServer': {
+            ...config['features.mockServer'],
+            ...config.apis[entrypointAlias]?.['features.mockServer'],
+          },
+          // TODO: merge everything else here
+        },
+        config.configFile,
+      )
+    : config;
+}
+
+export function transformConfig(rawConfig: DeprecatedRawConfig | RawConfig): RawConfig {
+  if ((rawConfig as RawConfig).apis && (rawConfig as DeprecatedRawConfig).apiDefinitions) {
+    throw new Error("Do not use 'apiDefinitions' field. Use 'apis' instead.\n");
+  }
+  if (
+    (rawConfig as RawConfig)['features.openapi'] &&
+    (rawConfig as DeprecatedRawConfig).referenceDocs
+  ) {
+    throw new Error("Do not use 'referenceDocs' field. Use 'features.openapi' instead.\n");
+  }
+  const { apiDefinitions, referenceDocs, ...rest } = rawConfig as DeprecatedRawConfig & RawConfig;
+  if (apiDefinitions) {
+    process.stderr.write(
+      `The ${yellow('apiDefinitions')} field is deprecated. Use ${green(
+        'apis',
+      )} instead. Read more about this change: https://redocly.com/docs/api-registry/guides/migration-guide-config-file/#changed-properties\n`,
+    );
+  }
+  if (referenceDocs) {
+    process.stderr.write(
+      `The ${yellow('referenceDocs')} field is deprecated. Use ${green(
+        'features.openapi',
+      )} instead. Read more about this change: https://redocly.com/docs/api-registry/guides/migration-guide-config-file/#changed-properties\n`,
+    );
+  }
+  return {
+    'features.openapi': referenceDocs,
+    apis: transformApiDefinitionsToApis(apiDefinitions),
+    ...rest,
+  };
+}
+
+export function getResolveConfig(resolve?: RawResolveConfig): ResolveConfig {
+  return {
+    http: {
+      headers: resolve?.http?.headers ?? [],
+      customFetch: undefined,
+    },
+  };
+}
+
+export function getUniquePlugins(plugins: Plugin[]): Plugin[] {
+  return plugins.reduce<Plugin[]>(
+    (acc, item) => (acc.some(({ id }) => id === item.id) ? acc : [...acc, item]),
+    [],
+  );
```

Comment:
This is hard to read and also it has O(n^2) complexity (which doesn't matter here because it is not hot path and n is going to be small but still):

```suggestion
  const seen = new Set();
  const results = [];
  for (const p of plugins) {
    if (!seen.has(p.id)) {
      results.push(p);
    } else {
      // we need to show some warning here maybe?
    }
    seen.add(p.id);
  }
  return results;
```

If we don't need to show a warning then it can be as simple as:

```suggestion
  return [ ...new Map(plugins.map(p => [p.id, p])).values() ];
```